### PR TITLE
config: implement config.Passwd.{Users.Groups}

### PR DESCRIPTION
**Pull request by vcaputo**

### CODE REVIEW

File: doc/config.md [hunk 1]
```diff
@@ -110,19 +110,21 @@ networkd:
...
-    gecos: "Foo Bar"
-    homedir: "/home/abc"
-    shell: "/bin/zsh"
-    no_create_home: false
-    no_user_group: false
-    system: false
-    no_log_init: false
     ssh_authorized_keys:
      - "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDEtomdD5MpmzEvLILIyn6l/TTrVOc2iRkeEgBjNEEFujkCOVMQrP7TFln+E2Ve0m5ngP+sImhYItpMpHOwRlBjUhFCzTZF8QJwUKyg2A59TV2uFEetosms7z4aW8lgQgm4m1ovi3G2R6BG3h01ogm3PC5YaAAkEbr0V0BxVN0rsTrq/dRNs2drLNw4giqJ5mBwzqXmepp7orJifyiKBueDQYsO367V7v9H797p3WmnFlg+T3LYiYfUQCkpxj/X+NYGbQBqsk5EXqAt/mi056HWu4esVJNwcIeiTdAWJ8/naKMAeEYoc8fJzbQ4rzIawRdAUk/QjxhVBbC4BxpGbsxr foouser@host"
+    password-hash: "$6$L5wKa16S$6qwF3RtAmPTlsDu/mVq53zzgL4htS0cw6GhFEDC9V6v6W7ydzxYN.dmJwlwBhMsi6vaNgkGvCRNwYhuFinA2w0"
```
Comment (crawford) [inline doc/config.md]: Can we use an underscore here instead?

File: src/exec/storage.go [hunk 2]
```diff
@@ -51,6 +51,16 @@ func (s storage) Run(config config.Config) bool {
 		return false
 	}
 
+	if err := s.createPasswd(config); err != nil {
+		s.Logger.Crit("failed to create users/groups: %v", err)
+		return false
+	}
+
+	if err := s.createFilesystemsFiles(config); err != nil {
```
Comment (crawford) [inline src/exec/storage.go]: Does this method belong in this commit?

File: doc/config.md (hunk 1 above)
Comment (crawford) [inline doc/config.md]: Actually, it looks like we intermix these throughout. We should standarized on '_', at least for JSON.

File: src/exec/storage.go (hunk 2 above)
Comment (vcaputo) [inline src/exec/storage.go]: Well, they're certainly related.

The creation of users/groups needs to happen before we create files which may be owned by those users, but after we create the filesystem the pwdb is stored on.

Comment (crawford) [inline src/exec/storage.go]: Doesn't this replace the existing method for writing files? I don't see any deleted lines in this PR.

File: doc/config.md [hunk 3]
```diff
@@ -110,19 +110,21 @@ networkd:
...
+    create:
+     primary_group: foouser
```
Comment (crawford) [inline doc/config.md]: Can you change underscores to space?

File: doc/config.md [hunk 4]
```diff
@@ -110,19 +110,21 @@ networkd:
...
+     shell: "/bin/zsh"
+     no_create_home: false
+     no_user_group: false
+     system: false
+     no_log_init: false
 
  groups:
   - name: foogroup
-    password: "$6$L5wKa16S$6qwF3RtAmPTlsDu/mVq53zzgL4htS0cw6GhFEDC9V6v6W7ydzxYN.dmJwlwBhMsi6vaNgkGvCRNwYhuFinA2w0"
+    password-hash: "$6$L5wKa16S$6qwF3RtAmPTlsDu/mVq53zzgL4htS0cw6GhFEDC9V6v6W7ydzxYN.dmJwlwBhMsi6vaNgkGvCRNwYhuFinA2w0"
```
Comment (crawford) [inline doc/config.md]: password hash:

File: src/config/group.go [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package config
+
+type Group struct {
+	Name         string `json:"name,omitempty"          yaml:"name"`
+	PasswordHash string `json:"password-hash,omitempty" yaml:"password-hash"`
```
Comment (crawford) [inline src/config/group.go]: passwordHash, password hash

File: src/exec/storage.go (hunk 2 above)
Comment (vcaputo) [inline src/exec/storage.go]: Sort of, I hadn't removed the createFiles call from createFilesystems.

File: src/exec/util/user_lookup.c [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+static int user_lookup_fn(user_lookup_ctxt_t *ctxt) {
+	char		buf[16 * 1024];
+	struct passwd	p, *pptr;
+
+	if(chroot(ctxt->root) == -1) {
+		goto out_err;
+	}
+
+	if(getpwnam_r(ctxt->name, &p, buf, sizeof(buf), &pptr) != 0 || !pptr) {
```
Comment (crawford) [inline src/exec/util/user_lookup.c]: Can `pptr` be NULL? I don't see anything in the manpage.

File: src/exec/util/user_lookup.c [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+#include "user_lookup.h"
+
+#define STACK_SIZE (64 * 1024)
+
+/* This is all a bit copy-and-pasty from update-ssh-keys/authorized_keys_d,
+ * TODO(vc): refactor authorized_keys_d a bit so external packages can reuse
+ * the pieces duplicated here.
+ */
+typedef struct user_lookup_ctxt {
+	void			*stack;
```
Comment (crawford) [inline src/exec/util/user_lookup.c]: Spaces for formatting please.

File: src/exec/util/user_lookup.c [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+					.root = root,
+					.name = name,
+					.res = res,
+					.ret = 0
+				};
+	int			pid, ret = 0;
+	sigset_t		allsigs, orig;
+
+	if(!(ctxt.stack = malloc(STACK_SIZE))) {
+		ret = -1;
```
Comment (crawford) [inline src/exec/util/user_lookup.c]: I think `return -1` here is a little easier to follow. Then you can drop the second label.

File: src/exec/util/user_lookup.c (hunk 6 above)
Comment (crawford) [inline src/exec/util/user_lookup.c]: getpwnam_r is stupid. Carry on.

File: src/exec/util/user_lookup.c (hunk 7 above)
Comment (vcaputo) [inline src/exec/util/user_lookup.c]: I align the variable names to the next tab stop so they're less likely to move should this code evolve changing any variables in this block.  If the names are kept tight, like one space over from the widest type, then they're more likely to require vertical realignment should any of their types change.

File: src/exec/util/user_lookup.c (hunk 8 above)
Comment (vcaputo) [inline src/exec/util/user_lookup.c]: if I have exit labels in a function, I tend to use them uniformly for various reasons.

**Comment by crawford** [comment]
L
	G
		T
	M

**Pull request merged**